src/main/scala/io/iohk/ethereum/blockchain/sync/RegularSync.scala

```diff
@@ -276,78 +259,58 @@ class RegularSync(
         responseMsgCode = BlockBodies.code)))
   }
 
-  def getOldBlocks(blockNumbers: Seq[BigInt]): List[Block] = blockNumbers match {
-    case Seq(blockNumber, tail @ _*) =>
-      blockchain.getBlockByNumber(blockNumber).map(_ :: getOldBlocks(tail)).getOrElse(Nil)
-    case Seq() =>
-      Nil
-  }
-
   private def handleBlockBodies(peer: Peer, m: Seq[BlockBody]) = {
     if (m.nonEmpty && headersQueue.nonEmpty) {
       val blocks = headersQueue.zip(m).map{ case (header, body) => Block(header, body) }
 
-      blockchain.getBlockHeaderByHash(blocks.head.header.parentHash)
-        .flatMap(b => blockchain.getTotalDifficultyByHash(b.hash)) match {
-        case Some(blockParentTd) =>
-          val (newBlocks, errorOpt) = processBlocks(blocks, blockParentTd)
-
-          if(newBlocks.nonEmpty)
-            log.debug(s"got new blocks up till block: ${newBlocks.last.block.header.number} " +
-              s"with hash ${Hex.toHexString(newBlocks.last.block.header.hash.toArray[Byte])}")
-
-          errorOpt match {
-            case Some(error) =>
-              val numberBlockFailed = blocks.head.header.number + newBlocks.length
-              resumeWithDifferentPeer(peer, reason = s"a block execution error: ${error.toString}, in block $numberBlockFailed")
-            case None =>
-              headersQueue = headersQueue.drop(blocks.length)
-              if (headersQueue.nonEmpty) {
-                val hashes = headersQueue.take(blockBodiesPerRequest).map(_.hash)
-                requestBlockBodies(peer, GetBlockBodies(hashes))
-              } else {
-                context.self ! ResumeRegularSync
-              }
-          }
-        case None =>
-          //TODO: Investigate if we can recover from this error (EC-165)
-          val parentHash = Hex.toHexString(blocks.head.header.parentHash.toArray)
-          log.error(s"No total difficulty for the latest block with number ${blocks.head.header.number - 1} (and hash $parentHash)")
-          shutdownAppFn()
+      @tailrec
+      def importBlocks(blocks: List[Block], importedBlocks: List[Block] = Nil): (List[Block], Option[BlockImportFailed]) =
```

Comment:
🤔

Reply:
I don't think it's a good idea. There is no logic related to importing blocks here, it just iterates over the blocks and calls `importBlocks` on them, the only logic is dealing with the import result (`BlockImportedToTop`, `ChainReorganised`, etc.). That's what is meant for `Ledger#importBlocks` - it should be a general use single entry point, and the callers should decide what to do with the result of a single block import. Different callers might want to handle it differently.

Reply:
sounds reasonable, but if we would add this as function to ledger, i think we would be able to test that logic in blockchain testsuite instead of just doing `blocksToProcess.foreach(ledger.importBlock)` do you think it would be valuable? (if not lets leave it as it is)

Reply:
In general there are imaginably different strategies on handling `BlockImportResult` for individual blocks, and therefore there cannot be a general case `importBlocks` function.

Even currently, usages of `importBlock` on sequences of blocks are different in `RegularSync` and `BlockchainSuite`. In the former, when handling `BlockBodies` message, we expect the blocks to form a single ordered sub-chain and we give up on execution on the first error. In the latter we are dealing with different sub-chains and `importBlocks` as defined `RegularSync` would simply not work.

Note that `importBlocks` is not even a private method. It's a nested function because it's applicable only to the case of `BlockBodies` message. The function itself is just a functional **loop** and nothing else.